File: Week 08/id_262/reverse-only-letters.py
class Solution:
    def reverseOnlyLetters(self, S: str) -> str:
        start, end = 0, len(S) - 1
        str_list = set("QWERTYUIOPLKJHGFDSAZXCVBNMqwertyuioplkjhgfdsazxcvbnm")
        s = [item for item in S]
        for _ in s:
            if start >= end:
                break
            if s[start] in str_list and s[end] in str_list:
                s[start], s[end] = s[end], s[start]
                start += 1
                end -= 1
            elif s[start] in str_list and s[end] not in str_list:
                end -= 1
            elif s[start] not in str_list and s[end] in str_list:
                start += 1
            else:
                start += 1
                end -= 1
        return "".join(s)

File: Week 08/id_237/reverseOnlyLetters.java
class Solution {
    public String reverseOnlyLetters(String S) {
        char[] chars = S.toCharArray();
        int i = 0;
        int j = chars.length - 1;
        while (i < j) {
            while (i < j && (chars[i] < 'A' || (chars[i] > 'Z' && chars[i] < 'a') || chars[i] > 'z')) i++;
            while (i < j && (chars[j] < 'A' || (chars[j] > 'Z' && chars[j] < 'a') || chars[j] > 'z')) j--;
            if(i < j) {
                chars[i] ^= chars[j];
                chars[j] ^= chars[i];
                chars[i++] ^= chars[j--];
            }
        }
        return String.valueOf(chars);
    }
}